Do not capture closed input channel in `Database.with_database`

Halbaroth opened the pull request:
We do not execute again `Database.with_database` if the database is loaded in main memory. In particular, all the channels open in this function are closed after executing `Driver.load_database`. Alas, we capture a closed channel in `index_of_string`. This commit fixes this bug.

The inversed index for strings is now loaded in main memory when we use Ancient.

Halbaroth committed:
Do not capture closed input channel in `Database.with_database`

We do not execute again `Database.with_database` if the database is
loaded in main memory. In particular, all the channels open in this
function are closed after executing `Driver.load_database`. Alas,
we capture a closed channel in `index_of_string`. This commit
fixes this bug.

The inversed index for strings is now loaded in main memory when we use
Ancient.

hgouraud commented:
tested this PR with base Wizard for Roglo. The server does not crash anymore, but the modification to an individual does not seem to be taken into account

Halbaroth commented:
You mean the individual does not change at all or the inverted index of strings is wrong?

a2line commented:
For the short test I did, the form data are stored as they should (no error, same process) but they are not shown until the server is restarted.

Halbaroth commented:
I need more information. I modified an individual as follows:
I changed his first name: Henri Aaron -> Henriette Aaron 

The change was immediately visible on both his (her? :p) page and searching pages.

So please, for the sake of god, tell me EXACTLY what you did.

hgouraud commented:
on the server next.roglo.eu, with commit 280e701bd, base wizard
```
henri@roglo6:~/github/hgouraud/geneweb$ git lgg
280e701bd (HEAD, origin/wip-roglo) Merge remote-tracking branch 'halbaroth/fix-index-load' into wip-roglo
3bdd6d744 Do not capture closed input channel in `Database.with_database`
5ab367a82 (geneweb/master) externalise main search input tooltip on welcome
b1d9166e2 add fallback and warning when book entry doesn’t exist
```
I tried to modify a note, a birth date, and just now a firstname.
None of them showed up !

you can try, you have wizard access to this base. Dy you remember your id:passwd?

gwd launched (as su) with /home/bin/start_gwdm.sh

Halbaroth commented:
Thanks :), I don't remember my passwd, could you send me a fresh one by mail?

hgouraud commented:
Could it be a conflict between data in memory and data in patches??

a2line commented:
I added a caracter to the occupation field on two differents individuals.

a2line commented:
> Could it be a conflict between data in memory and data in patches??

it's what I thought yesterday, if patched data change they need to be reloaded in memory?

Halbaroth committed:
Ensure that workers read again patches before each request

This commit is a hotfix. Before introducing worker pool and refactoring
`Driver` API for database memory, patches on the disk were read
before each request. I believed that we could reuse `db_disk` value
between requests. I was wrong! The database value contains several
closures capturing copies of patches in memory. We must read patches
on the disk to ensure that workers remain synchronized.

Halbaroth committed:
Do not load inverted index in memory

This cache is useless because we must read patches on the disk for each
request.

Halbaroth commented:
I fixed the second bug. The issue comes from a bad interaction between worker pools, in-memory database and patch system. When I introduced `Driver.load_database`, I hoped that we could reuse the previous base value, but I was wrong. https://github.com/geneweb/geneweb/pull/2292/commits/0c21dc22e8fb727278f4f88434246d01edee158e contains a hotfix. 

We cannot load inverted index for strings in memory because we need to read patches on disk for each request.
